torch/linalg/__init__.py

```diff
@@ -88,6 +88,49 @@
 Alias of :func:`torch.det`.
 """)
 
+matrix_rank = _add_docstr(_linalg.linalg_matrix_rank, r"""
+matrix_rank(input, tol=None, hermitian=False) -> Tensor
+
+Computes the numerical rank of a matrix :attr:`input`,
+or of each matrix in a batched :attr:`input`. The method to compute the
+matrix rank is done using SVD by default. If :attr:`hermitian` is ``True``,
+then :attr:`input` is assumed to be Hermitian (symmetric if real-valued),
+and the computation of the rank is done by obtaining the eigenvalues.
+
+:attr:`tol` is the threshold below which the singular values (or the eigenvalues
+when :attr:`hermitian` is ``True``) are considered to be 0. If :attr:`tol` is not
+specified, :attr:`tol` is set to ``S.max() * max(S.size()) * eps`` where `S` is the
+singular values (or the eigenvalues when :attr:`hermitian` is ``True``), and ``eps``
+is the epsilon value for the datatype of :attr:`input`.
+
+Supports input of ``float``, ``double``, ``cfloat`` and ``cdouble`` datatypes.
+
+Args:
+    input (Tensor): the input matrix of size :math:`(m, n)` or the batch of matrices of size :math:`(*, m, n)`
+                    where `*` is one or more batch dimensions.
+    tol (float, optional): the tolerance value. Default: ``None``
+    hermitian(bool, optional): indicates whether :attr:`input` is Hermitian. Default: ``False``
+
+Examples::
+
+    >>> a = torch.eye(10)
```

Comment:
These examples are fine, but in this PR (or a future PR) we may want to show the intermediate steps of each computation (Hermitian or not) in the examples.

Reply:
I added examples that use `tol` and `hermitian` arguments:
```python
    >>> a = torch.randn(2, 4, 3, 3, dtype=torch.complex64)
    >>> torch.linalg.matrix_rank(a)
    tensor([[3, 3, 3, 3],
            [3, 3, 3, 3]])
    >>> torch.linalg.matrix_rank(a, hermitian=True)
    tensor([[3, 3, 3, 3],
            [3, 3, 3, 3]])
    >>> torch.linalg.matrix_rank(a, tol=1.0)
    tensor([[3, 2, 2, 2],
            [1, 2, 1, 2]])
    >>> torch.linalg.matrix_rank(a, tol=1.0, hermitian=True)
    tensor([[2, 2, 2, 1],
            [1, 2, 2, 2]])
```